File: src/main/java/io/myosotisdev/utopianism/util/DoubleTemplate.java
package io.myosotisdev.utopianism.util;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

public class DoubleTemplate implements INumberTemplate<Double>
{
    private double base;
    private final boolean useProportion;
    private final boolean useValues;
    private Map<String, ValueDouble> proportion;
    private Map<String, ValueDouble> values;
    private Runnable onValueChanged;

    private double latest = -1;

    public DoubleTemplate(double base)
    {
        this(base, true, true);
    }

    public DoubleTemplate(double base, Runnable onValueChanged)
    {
        this(base, true, true);
        this.onValueChanged = onValueChanged;
    }

    public DoubleTemplate(double base, boolean useProportion, boolean useValues)
    {
        this.base = base;
        this.useProportion = useProportion;
        this.useValues = useValues;
        this.proportion = new HashMap<>();
        this.values = new HashMap<>();
    }

    public void setOnValueChanged(Runnable runnable)
    {
        this.onValueChanged = runnable;
    }

    @Override
    public Double requestValue()
    {
        latest = base
                + (useProportion ? new BigDecimal(base * (proportion.values().stream().mapToDouble(value -> value.doubleValue()).reduce(0, Double::sum))).intValue() : 0)
                + (useValues ? values.values().stream().mapToDouble(value -> value.doubleValue()).reduce(0, Double::sum) : 0);
        if (onValueChanged != null)
        {
            onValueChanged.run();
        }
        return getValue();
    }

    @Override
    public void addProportion(String key, double percentage, int duration)
    {

    }

    @Override
    public void addValue(String key, int value, int duration)
    {

    }

    @Override
    public Double getValue()
    {
        return latest;
    }
}

File: src/main/java/io/myosotisdev/utopianism/util/DoubleTemplate.kt
package io.myosotisdev.utopianism.util

import java.math.BigDecimal
import java.util.concurrent.ConcurrentHashMap

class DoubleTemplate(private val base: Double, private val useProportion: Boolean = true, private val useValues: Boolean = true) : INumberTemplate<Double>
{
    private val proportion: ConcurrentHashMap<String, ValueDouble> = ConcurrentHashMap()
    private val values: ConcurrentHashMap<String, ValueDouble> = ConcurrentHashMap()
    private val updated = false
    private var latest: Double = 0.0

    override fun addProportion(key: String?, percentage: Double, duration: Int)
    {
    }

    override fun addValue(key: String?, value: Double, duration: Int)
    {
    }

    override fun removeProportion(key: String?)
    {
        TODO("Not yet implemented")
    }

    override fun removeValue(key: String?)
    {
        TODO("Not yet implemented")
    }

    override fun value(): Double
    {
        return if (updated) (base
                             + (if (useProportion) BigDecimal(base * proportion.values.stream()
                .mapToDouble { obj: ValueDouble -> obj.value }
                .reduce(0.0) { a: Double, b: Double -> java.lang.Double.sum(a, b) }).toDouble()
        else 0.0)
                             + if (useValues) values.values.stream()
                .mapToDouble { obj: ValueDouble -> obj.value }
                .reduce(0.0) { a: Double, b: Double -> a + b }
                             else 0.0).also { latest = it }
        else latest
    }
}